### PR TITLE
Enforce consistency when accessing array with non-integer index values

### PR DESCRIPTION
**Pull request by davidegrohmann**
The problem was that when using non integer indexes for accessing
arrays we fail with a semantic exception when the index type is known
in advance of not being an integer (e.g., when using a literal) whilst
we success at run time by coarcing the value to an integer (e.g., when
using parameters).  Such behaviour is confusing and upredictable,
hence this change will make both cases to fail, the first while
semantic checking and the second at runtime.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/CollectionIndexTest.scala [hunk 1]
```diff
@@ -55,6 +55,22 @@ class CollectionIndexTest extends CypherFunSuite {
...
+    a [CypherTypeException] should be thrownBy idx(1.0d)
+    a [CypherTypeException] should be thrownBy idx("bad value")
+  }
+
+  test("should fail when too bigger values are used to access the array") {
+    implicit val collection = Literal(Seq(1, 2, 3, 4))
+
+    val index = Int.MaxValue + 1L
+
+    a [InvalidArgumentException] should be thrownBy idx(index)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/CollectionIndexTest.scala]: an

**Comment by craigtaverner** [comment]
Should this behavior change be fixed as far back as 2.2? While it is a bugfix, it is also a behavior change, which we'd normally be cautious about back-porting.

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/CollectionIndexTest.scala [hunk 2]
```diff
@@ -55,6 +55,22 @@ class CollectionIndexTest extends CypherFunSuite {
...
 
+  test("should fail when not integer values are passed") {
+    implicit val collection = Literal(Seq(1, 2, 3, 4))
+
+    a [CypherTypeException] should be thrownBy idx(1.0f)
+    a [CypherTypeException] should be thrownBy idx(1.0d)
+    a [CypherTypeException] should be thrownBy idx("bad value")
+  }
+
+  test("should fail when too bigger values are used to access the array") {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/CollectionIndexTest.scala]: `too big` or `too large`

**Comment by pontusmelke** [comment]
+1 on what @craigtaverner is saying

**Comment by davidegrohmann** [comment]
@craigtaverner If this is considered a bug, yes I really think it has to be fixed as far back as 2.2. If considered a behavioral change I am happy to make it in 3.0 only. Unfortunately there is no agreement on this.

**Comment by Mats-SX** [comment]
@davidegrohmann The CLG decision is to go through with the fix you've implemented (throw exception in both cases), but to target 3.0. Could you rebase this, please?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherTypeAcceptanceTest.scala [hunk 3]
```diff
@@ -70,4 +70,9 @@ class CypherTypeAcceptanceTest extends ExecutionEngineFunSuite {
 
     result should be(empty)
   }
+
+  test("should be consistent in failing either statically or at runtime when trying to access an array with a non-integer index") {
+    a [SyntaxException] should be thrownBy execute("WITH [1,2,3,4,5] AS array, 3.14 AS idx RETURN array[idx]")
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherTypeAcceptanceTest.scala:75]: Should this really be `SyntaxException`? Isn't it rather a semantic error?

Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherTypeAcceptanceTest.scala:75]: All semantic errors are mapped to `SyntaxException`s. Not that there is no `SemanticError` public exception. We might want to change the current situation, but this comment is unrelated to this specific fix.

Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherTypeAcceptanceTest.scala:75]: All right!
